index.ts

```diff
@@ -0,0 +1,552 @@
+  gravity?: boolean; // Prevented from being executed or selected during Gravity's effect.
+  heal?: boolean; // Prevented from being executed or selected during Heal Block's effect.
+  mirror?: boolean; // Can be copied by Mirror Move.
+  mystery?: boolean; // Unknown effect.
+  nonsky?: boolean; // Prevented from being executed or selected in a Sky Battle.
+  powder?: boolean; // Has no effect on Pokemon which are Grass-type, have the Ability Overcoat, or hold Safety Goggles.
+  protect?: boolean; // Blocked by Detect, Protect, Spiky Shield, and if not a Status move, King's Shield.
+  pulse?: boolean; // Power is multiplied by 1.5 when used by a Pokemon with the Ability Mega Launcher.
+  punch?: boolean; // Power is multiplied by 1.2 when used by a Pokemon with the Ability Iron Fist.
+  recharge?: boolean; // If this move is successful, the user must recharge on the following turn and cannot make a move.
+  reflectable?: boolean; // Bounced back to the original user by Magic Coat or the Ability Magic Bounce.
+  snatch?: boolean; // Can be stolen from the original user and instead used by another Pokemon using Snatch.
+  sound?: boolean; // Has no effect on Pokemon with the Ability Soundproof.
+}
+
+type MoveCategory: 'Physical' | 'Special' | 'Status';
+
+type MoveTarget =
+  // single-target
+  'normal'|'any'|'adjacentAlly'|'adjacentFoe'|'adjacentAllyOrSelf'|
+  // single-target, automatic
+  'self'|'randomNormal'|
+  // spread
+  'allAdjacent'|'allAdjacentFoes'|
+  // side and field
+  'allySide'|'foeSide'|'all';
+
+interface SelfEffect {
+  volatileStatus?: ID;
+  boosts?: Partial<BoostsTable>;
+}
+
+interface SecondaryEffect {
+  chance?: number;
+  boosts?: Partial<BoostsTable>;
+  self?: SelfEffect;
+  status?: StatusName;
+  volatileStatus?: ID;
+}
+
+type ContestType = 'Cool' | 'Beautiful' | 'Cute' | 'Clever' | 'Tough';
+
+interface CommonMove<TypeT, ItemT, MoveT> {
+  accuracy: number | true;
+  basePower: number;
+  flags: MoveFlags;
+  pp: number;
+  target: MoveTarget;
+  type: TypeT;
+
+  alwaysCrit?: boolean;
+  alwaysHit?: boolean;
+  basePowerModifier?: number;
+  boosts?: Partial<BoostsTable> | false
+  bypassesProtect?: boolean;
+  category: MoveCategory;
+  contestType?: ContestType;
+  critModifier?: number;
+  critRatio?: number;
+  damage?: number | 'level';
+  defensiveCategory?: MoveCategory;
+  drain?: [number, number]; // fraction
+  forceSTAB?: boolean;
+  forceSwitch?: boolean;
+  givesHealth?: boolean;
+  hasCustomRecoil?: boolean;
+  hasPriority?: boolean;
+  hasSecondaryEffect?: boolean;
+  heal?: [number, number]; // fraction
+  ignoreAbility?: boolean;
+  ignoreAccuracy?: boolean;
+  ignoreDefensive?: boolean;
+  ignoreEvasion?: boolean;
+  ignoreImmunity?: boolean | TypeT;
+  ignoreNegativeOffensive?: boolean;
+  ignoreOffensive?: boolean;
+  ignorePositiveDefensive?: boolean;
+  ignorePositiveEvasion?: boolean;
+  ignoresBurn?: boolean;
+  ignoresDefenseBoosts?: boolean;
```

Comment:
My eyes are glazing over checking these, are they moveflags? should they be combined into `ignore`?

Reply:
I really don't know. If we no longer have the 'MoveFlags are the literally bit flags from the device' distinction, it seems we'd want all booleans to be flags IMO. If the logic is 'theyre only set by N moves and thus are not generally useful enough to be flags' it seems to imply that they might instead be being used for an implementation detail? @Marty-D @Zarel

Reply:
I think they should be move flags. I'm resistant to hardcoding because it causes trouble for OMs.